### PR TITLE
MultiRC

### PR DESCRIPTION
**Pull request by W4ngatang**
Add MultiRC task, including:

- task definition
- task specific module and forward function
- prediction generation

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -246,16 +246,16 @@ def evaluate_and_write(args, model, tasks, splits_to_write):
...
         evaluate.write_preds(tasks, te_preds, args.run_dir, 'test',
                              strict_glue_format=args.write_strict_glue_format)
     run_name = args.get("run_name", os.path.basename(args.run_dir))
 
     results_tsv = os.path.join(args.exp_dir, "results.tsv")
     log.info("Writing results for split 'val' to %s", results_tsv)
     evaluate.write_results(val_results, results_tsv, run_name=run_name)
-
+    if 'test' in splits_to_write:
+        evaluate.write_results(te_results, results_tsv, run_name=run_name)
```
Comment (sleepinyourhat) [inline main.py]: Why do we only write *results* if splits_to_write is set?

File: src/evaluate.py [hunk 2]
```diff
@@ -148,6 +150,9 @@ def write_preds(tasks: Iterable[tasks_module.Task], all_preds, pred_dir, split_n
         elif isinstance(task, CommitmentTask):
```
Comment (sleepinyourhat) [inline src/evaluate.py]: We should really move these into the task objects. Mind making an issue?

File: src/evaluate.py [hunk 3]
```diff
@@ -168,7 +173,8 @@ def write_preds(tasks: Iterable[tasks_module.Task], all_preds, pred_dir, split_n
                  'sts-b': 'STS-B',
                  'wnli': 'WNLI'}
 
-SUPERGLUE_NAME_MAP = {"commitbank": 'CB'
```
Comment (sleepinyourhat) [inline src/evaluate.py]: Add a comment (here and for GLUE) explaining what these are for.

File: src/evaluate.py [hunk 4]
```diff
@@ -38,7 +40,7 @@ def parse_write_preds_arg(write_preds_arg: str) -> List[str]:
 def evaluate(model, tasks: Sequence[tasks_module.Task], batch_size: int,
              cuda_device: int, split="val") -> Tuple[Dict, pd.DataFrame]:
     '''Evaluate on a dataset'''
-    FIELDS_TO_EXPORT = ['idx', 'sent1_str', 'sent2_str', 'labels']
+    FIELDS_TO_EXPORT = ['idx', 'sent1_str', 'sent2_str', 'labels', 'qst_idx', 'ans_idx']
```
Comment (sleepinyourhat) [inline src/evaluate.py]: I assume you've confirmed that this doesn't break other tasks.

File: src/models.py [hunk 5]
```diff
@@ -686,6 +690,15 @@ def build_decoder(task, d_inp, vocab, embedder, args):
     hid2voc = nn.Linear(args.s2s['d_hid_dec'], args.max_word_v_size)
     return decoder, hid2voc
 
+def build_qa_module(task, d_inp, use_bert, params):
+    ''' '''
+    pool_type = "first" if use_bert else "max"
+    #d_inp = d_inp if not use_bert else d_inp
```
Comment (sleepinyourhat) [inline src/models.py]: Cut?

File: src/models.py [hunk 6]
```diff
@@ -686,6 +690,15 @@ def build_decoder(task, d_inp, vocab, embedder, args):
     hid2voc = nn.Linear(args.s2s['d_hid_dec'], args.max_word_v_size)
     return decoder, hid2voc
```
Comment (sleepinyourhat) [inline src/models.py]: autopep8 to make sure you get the whitespace right.

File: src/models.py [hunk 7]
```diff
@@ -686,6 +690,15 @@ def build_decoder(task, d_inp, vocab, embedder, args):
...
 
+def build_qa_module(task, d_inp, use_bert, params):
+    ''' '''
```
Comment (sleepinyourhat) [inline src/models.py]: Fill in?

File: src/models.py [hunk 8]
```diff
@@ -1237,6 +1252,40 @@ def _grounded_ranking_bce_forward(self, batch, task, predict):
...
 
+    def _multiple_choice_reading_comprehension_forward(self, batch, task, predict):
+        ''' Forward call for multiple choice (selecting from a fixed set of answers)
+        reading comprehension (have a supporting paragraph).
+
+        Batch has a tensor of shape (n_questions, n_answers, n_tokens)
+        '''
+        out = {}
+        classifier = self._get_classifier(task)
+        if self.use_bert:
```
Comment (sleepinyourhat) [inline src/models.py]: Add a brief comment for the if and else cases.

File: src/tasks/qa.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+from ..utils.utils import truncate
+
+from .tasks import Task, sentence_to_text_field
+from .tasks import sentence_to_text_field, atomic_tokenize
+from .tasks import UNK_TOK_ALLENNLP, UNK_TOK_ATOMIC
+from .registry import register_task
+
+@register_task('multirc', rel_path='MultiRC/')
+class MultiRCTask(Task):
+    '''Multiple sentence reading comprehension Task'''
```
Comment (sleepinyourhat) [inline src/tasks/qa.py]: Multi-Sentence Reading Comprehension
https://cogcomp.org/multirc/

File: src/models.py [hunk 10]
```diff
@@ -1237,6 +1252,40 @@ def _grounded_ranking_bce_forward(self, batch, task, predict):
...
+            inp = batch["para_quest_ans"]
+            ex_embs, ex_mask = self.sent_encoder(inp, task)
+            logits = classifier(ex_embs, ex_mask)
+            out['n_exs'] = inp["bert_wpm_pretokenized"].size(0)
+        else:
```
Comment (sleepinyourhat) [inline src/models.py]: Since BERT is the only supported tokenizer, add a note or assert or something here.

Comment (sleepinyourhat) [reply src/models.py]: Also, how do we get CBOW results in that case?

File: src/tasks/qa.py [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+                    ans = answer["text"]
+                    label = int(answer["isAnswer"]) if "isAnswer" in answer else 0
+                    yield _make_instance(para, question, ans, label, par_idx, qst_idx, ans_idx)
+
+
+    def count_examples(self):
+        ''' Compute here b/c we're streaming the sentences. '''
+        example_counts = {}
+        for split, split_path in self.files_by_split.items():
+            #example_counts[split] = sum(len(ex["paragraph"]["questions"]) for ex in \
```
Comment (sleepinyourhat) [inline src/tasks/qa.py]: Cut?

File: src/models.py (hunk 10 above)
Comment (W4ngatang) [reply src/models.py]: It works with arbitrary tokenization, this branch is just if we're using BERT.

File: src/evaluate.py (hunk 4 above)
Comment (W4ngatang) [reply src/evaluate.py]: Yea, there's a line shortly below that's `for field in FIELDS_TO_EXPORT: if field in batch`

File: main.py (hunk 1 above)
Comment (W4ngatang) [reply main.py]: Not sure if I understand, predictions are written in `evaluate.write_preds(task, te_preds, ...)` above.

File: src/evaluate.py (hunk 2 above)
Comment (W4ngatang) [reply src/evaluate.py]: done in #635

File: src/tasks/qa.py [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+        self.val_metric = "%s_avg" % self.name
+        self.val_metric_decreases = False
+        self.max_seq_len = max_seq_len
+        self.files_by_split = {"train": os.path.join(path, "train.jsonl"),
+                               "val": os.path.join(path, "val.jsonl"),
+                               "test": os.path.join(path, "test.jsonl")}
+
+    def tokenizer_is_supported(self, tokenizer_name):
+        ''' For now, only support BERT '''
+        return "bert" in tokenizer_name
```
Comment (sleepinyourhat) [inline src/tasks/qa.py]: What about this?

Comment (W4ngatang) [reply src/tasks/qa.py]: Oops, this is deprecated and I'll delete it.

File: src/evaluate.py [hunk 13]
```diff
@@ -150,6 +152,10 @@ def write_preds(tasks: Iterable[tasks_module.Task], all_preds, pred_dir, split_n
         elif isinstance(task, COPATask):
             _write_copa_preds(task, preds_df, pred_dir, split_name,
                               strict_glue_format=strict_glue_format)
+
```
Comment (sleepinyourhat) [inline src/evaluate.py]: Delete extra line if easy.

File: src/evaluate.py [hunk 14]
```diff
@@ -264,6 +273,27 @@ def _write_copa_preds(task, preds_df: pd.DataFrame,
...
+        if strict_glue_format:
+            qst_ans_d = defaultdict(list)
+            for row_idx, row in preds_df.iterrows():
+                ans_d = {"idx": int(row["ans_idx"]), "label": int(row["preds"])}
+                qst_ans_d[int(row["qst_idx"])].append(ans_d)
+
+            for qst_idx, answers in qst_ans_d.items():
+                out_d = {"idx": qst_idx, "answers": answers}
+                preds_fh.write("{0}\n".format(json.dumps(out_d)))
+
```
Comment (sleepinyourhat) [inline src/evaluate.py]: Delete extra lines if easy. Here and above.

**Pull request merged**